rust/src/dataset/fragment.rs

```diff
@@ -66,35 +68,27 @@ impl FileFragment {
     pub async fn create(
         dataset_uri: &str,
         id: usize,
-        reader: &mut dyn RecordBatchReader,
+        reader: impl RecordBatchReader + Send + 'static,
         params: Option<WriteParams>,
     ) -> Result<Fragment> {
         let params = params.unwrap_or_default();
 
-        let schema = Schema::try_from(reader.schema().as_ref())?;
+        let reader = Box::new(reader);
+        let (stream, schema) = reader_to_stream(reader)?;
+
         let (object_store, base_path) = ObjectStore::from_uri(dataset_uri).await?;
         let filename = format!("{}.lance", Uuid::new_v4());
         let fragment = Fragment::with_file(id as u64, &filename, &schema);
 
         let full_path = base_path.child(DATA_DIR).child(filename.clone());
 
         let mut writer = FileWriter::try_new(&object_store, &full_path, schema.clone()).await?;
-        let mut batches: Vec<RecordBatch> = Vec::new();
-        let mut num_rows: usize = 0;
-        for rst in reader {
-            let batch: RecordBatch = rst?; // TODO: close writer on Error?
-            batches.push(batch.clone());
-            num_rows += batch.num_rows();
-            if num_rows >= params.max_rows_per_group {
-                writer.write(&batches).await?;
-                batches = Vec::new();
-                num_rows = 0;
-            }
-        }
 
-        if num_rows > 0 {
-            writer.write(&batches).await?;
-        };
+        let mut buffered_reader = chunk_stream(stream, params.max_rows_per_file);
```

Comment:
Should this be `max_rows_per_group` instead?

Reply:
Oh you are right, this is a typo.

Reply:
Fixed.